### ISSUE TITLE
Bug: (17.0.0-rc.1) lazy is eager in dev mode

### ISSUE THREAD

**Issue by nstepien** [facebook/react#19748]
React version: 17.0.0-rc.1

## Steps To Reproduce

1. Create elements out of `lazy` components, use them conditionally.
2. Check if the `lazy` components are dynamically imported.

Link to code example: --

`main.jsx`:
```jsx
import React, { lazy, Suspense, useState } from 'react';
import { render } from 'react-dom';

const Checked = lazy(() => import('./Checked'));
const Checked2 = lazy(() => import('./Checked2'));
const Unchecked = lazy(() => import('./Unchecked'));
const Unchecked2 = lazy(() => import('./Unchecked2'));

function App() {
  const [checked, setChecked] = useState(false);

  const checkedElement = <Checked />;
  const uncheckedElement = <Unchecked />;

  return (
    <>
      <label>
        <input
          type="checkbox"
          checked={checked}
          onChange={e => setChecked(e.target.checked)}
        />
        Toggle me
      </label>

      <hr />

      <Suspense fallback="loading...">
        Checked? {checked ? checkedElement : uncheckedElement}
      </Suspense>

      <hr />

      <Suspense fallback="loading...">
        Checked? {checked ? <Checked2 /> : <Unchecked2 />}
      </Suspense>
    </>
  );
}

render(<App />, document.getElementById('app'));
```
`Checked.jsx`:
```jsx
export default function() {
  return 'Checked';
}
```
`Checked2.jsx`:
```jsx
export default function() {
  return 'Checked 2';
}
```
`Unchecked.jsx`:
```jsx
export default function() {
  return 'Unchecked';
}
```
`Unchecked2.jsx`:
```jsx
export default function() {
  return 'Unchecked 2';
}
```

## The current behavior
The `Checked.jsx` chunk is imported eagerly, even though the component isn't immediately rendered.
![image](https://user-images.githubusercontent.com/567105/91972475-b039e580-ed12-11ea-96a4-8c7dc6ee5388.png)

This doesn't happen in prod mode, or with React 16, or with components that are not `createElement`-ed eagerly.

This affects react-router, as it uses this pattern:
```jsx
<Router>
  <Suspense fallback={<Loading />}>
    <Switch>
      <Route exact path="/">
        <Home />
      </Route>
      <Route exact path="/demo">
        <Demo />
      </Route>
      <Route>
        <NotFound />
      </Route>
    </Switch>
  </Suspense>
</Router>
```

AFAICT it doesn't break anything, but it is unexpected/confusing.

## The expected behavior
`lazy` components that are not rendered should not be dynamically imported.

**Comment by nstepien** [comment]
Shorter test case:
```jsx
import React, { lazy } from 'react';

const Checked = lazy(() => import('./Checked'));

<Checked />;
```
This is enough to trigger the dynamic import in dev mode.

**Comment by eps1lon** [comment]
Behavior changed between 16.13.1 and the first `next` release after `16.13.1`.

- [16.13.1 repro](https://codesandbox.io/s/react-16131-createelement-for-lazy-knn1q?file=/src/index.js)
![Screenshot from 2020-09-03 14-22-26](https://user-images.githubusercontent.com/12292047/92114329-ee0d3b80-edf0-11ea-9afc-504cea77c51d.png)

- [0.0.0-d7382b6c4 repro](https://codesandbox.io/s/react000-d7382b6c4-createelement-for-lazy-vsq0i?file=/src/index.js)
  
  ![Screenshot from 2020-09-03 14-22-01](https://user-images.githubusercontent.com/12292047/92114337-f06f9580-edf0-11ea-9e52-c5d42d325888.png)

[Diff between 16.13.1 and 0.0.0-d7382b6c4](https://github.com/facebook/react/compare/v16.13.1...v0.0.0-d7382b6c4)

**Comment by gaearon** [comment]
This is triggered by `getComponentName` in `validatePropTypes`. I think this doesn't really make sense to call but we'd need to see why this change to `getComponentName` was made. @sebmarkbage?

<img width="398" alt="Screenshot 2020-09-03 at 13 56 14" src="https://user-images.githubusercontent.com/810438/92117590-45a9a800-eded-11ea-8195-301390c7c42e.png">
<img width="367" alt="Screenshot 2020-09-03 at 13 56 18" src="https://user-images.githubusercontent.com/810438/92117593-46dad500-eded-11ea-8dd5-1734fcbf1608.png">

**Comment by sebmarkbage** [comment]
The suspense protocol is just a function that may throw. That allows for low overhead abstractions to be wrapped. This change was to allow the init function to encapsulate the actual suspense mechanism and that data structure.

Unfortunately there’s no way to query the status of something in the suspense protocol without triggering the fetch.

Same as you can’t do:

```
var data = readData(...);
validateData(data);
```

Without triggering a fetch.

The solution might be to move validation of lazy until it’s rendered instead of in createElement.

Also, sigh PropTypes.

**Comment by nstepien** [comment]
Should the `Unconfirmed` label be removed?

**Comment by gaearon** [comment]
https://github.com/facebook/react/pull/19871

**Comment by gaearon** [comment]
Should be fixed in `17.0.0-rc.2`. Please verify.

**Comment by nstepien** [comment]
@gaearon Works great now, thanks!
Looking forward to the 17.0.0 release.
